src/main/java/org/mule/extension/file/internal/source/DirectoryListener.java

```diff
@@ -0,0 +1,491 @@
+    filesBeingProcessingObjectStore = objectStoreManager.getOrCreateObjectStore("file-listener:" + directory,
+                                                                                ObjectStoreSettings.builder()
+                                                                                    .persistent(false)
+                                                                                    .maxEntries(1000)
+                                                                                    .entryTtl(60000L)
+                                                                                    .expirationInterval(20000L)
+                                                                                    .build());
+
+    if (watermarkMode != DISABLED) {
+      watermarkObjectStore = objectStoreManager.getOrCreateObjectStore("file-listener-watermark:" + directory,
+                                                                       unmanagedPersistent());
+    }
+
+    long freq = poolingFrequencyTimeUnit.toMillis(pollingFrequency);
+    listenerExecutor.scheduleAtFixedRate(() -> poll(sourceCallback), freq, freq, MILLISECONDS);
+    started = true;
+    stopRequested.set(false);
+  }
+
+  @OnSuccess
+  public void onSuccess(@ParameterGroup(name = POST_PROCESSING_GROUP_NAME) PostActionGroup postAction,
+                        SourceCallbackContext ctx) {
+    postAction(postAction, ctx);
+  }
+
+  @OnError
+  public void onError(@ParameterGroup(name = POST_PROCESSING_GROUP_NAME) PostActionGroup postAction,
+                      SourceCallbackContext ctx) {
+    if (postAction.isApplyPostActionWhenFailed()) {
+      postAction(postAction, ctx);
+    }
+  }
+
+  @OnTerminate
+  public void onTerminate(SourceCallbackContext ctx) {
+    ctx.<Runnable>getVariable(FILE_RELEASER_VAR).ifPresent(Runnable::run);
+  }
+
+  private void poll(SourceCallback<InputStream, FileAttributes> sourceCallback) {
+    if (isRequestedToStop()) {
+      return;
+    }
+
+    try {
+      java.util.Optional<LocalDateTime> watermark = getWatermark();
+      PollWalker walker = new PollWalker(watermark, sourceCallback);
+      walkFileTree(directoryPath, EnumSet.of(FOLLOW_LINKS), MAX_VALUE, walker);
+
+      walker.getUpdatedWatermark().ifPresent(this::updateWatermark);
+
+    } catch (Exception e) {
+      LOGGER.error(format("Found exception trying to poll directory '%s'. Will try again on the next poll. ",
+                          directoryPath.toString(), e.getMessage()),
+                   e);
+    }
+  }
+
+  private FileVisitResult processFile(Path path, LocalFileAttributes attributes, Lock lock,
+                                      SourceCallback<InputStream, FileAttributes> callback) {
+    String fullPath = path.toString();
+    try {
+      if (filesBeingProcessingObjectStore.contains(fullPath)) {
+        LOGGER.debug("Polled file '{}', but skipping it since it is already being processed in another thread or node", fullPath);
```

Comment:
isEnabled

Reply:
no need to do is enabled here because i'm using slf4j interpolation, so no string will be built unnecesarily